assistant/assistant.go

```diff
@@ -0,0 +1,186 @@
+}
+
+/*
+AssistantConfig allows us to configure where to connect to and other params
+for the agent.
+*/
+type AssistantConfig struct {
+	Host          string
+	Path          string
+	RetryInterval int
+}
+
+func (astConfig *AssistantConfig) fillDefaults() {
+	if astConfig.Host == "" {
+		astConfig.Host = "localhost:8080"
+	}
+	if astConfig.Path == "" {
+		astConfig.Path = "/echo"
+	}
+	if astConfig.RetryInterval == 0 {
+		astConfig.RetryInterval = 15
+	}
+}
+
+// NewAssistant returns an Assistant for the given Config.
+func NewAssistant(ctx context.Context, config *AssistantConfig, agent *agent.Agent) (*Assistant, error) {
+	config.fillDefaults()
+	var addr = flag.String("addr", config.Host, "http service address")
+	u := url.URL{Scheme: "ws", Host: *addr, Path: config.Path}
+
+	header := http.Header{}
+
+	if v, exists := os.LookupEnv("INFLUX_TOKEN"); exists {
+		header.Add("Authorization", "Token "+v)
+	} else {
+		return nil, fmt.Errorf("Influx authorization token not found, please set in env")
+	}
+
+	// creates a new websocket connection
+	log.Printf("D! [assistant] Attempting connection to [%s]", config.Host)
+	ws, _, err := websocket.DefaultDialer.Dial(u.String(), header)
+	for err != nil { // on error, retry connection again
+		log.Printf("E! [assistant] Failed to connect to [%s], retrying in %ds, "+
+			"error was '%s'", config.Host, config.RetryInterval, err)
+
+		err := internal.SleepContext(ctx, time.Duration(config.RetryInterval)*time.Second)
+		if err != nil {
+			return nil, err
+		}
+
+		ws, _, err = websocket.DefaultDialer.Dial(u.String(), header)
+	}
+	log.Printf("D! [assistant] Successfully connected to %s", config.Host)
+
+	a := &Assistant{
+		config:     config,
+		connection: ws,
+		ctx:        ctx,
+		done:       make(chan bool),
+		agent:      agent,
+	}
+
+	go a.listenToServer()
+
+	return a, nil
+}
+
+// Stop is used to clean up active connection and all channels
+func (assistant *Assistant) Stop() {
+	assistant.connection.Close()
+	assistant.done <- true
+	close(assistant.done)
+}
+
+type plugin struct {
+	Name   string
+	Type   string
+	Config map[string]interface{}
+}
+
+type Request struct {
+	Operation string
+	Uuid      string
+	Plugin    plugin
+}
+
+type Response struct {
+	Status string
+	Uuid   string
+	Data   interface{}
+}
+
+// WriteToServer is used as a helper function to write status responses to server.
+func (assistant *Assistant) WriteToServer(payload Response) error {
+	/*
+		TODO: Write a dedicated struct for the input of WriteToServer.
+		Refer to the design doc for implementation.
+		https://docs.google.com/document/d/1pnXrWgXCvlpe5tB3YAlOtO6rsvBxoV7DMPHL82KpCok/edit?usp=sharing
+	*/
+
+	err := assistant.connection.WriteJSON(payload)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (assistant *Assistant) checkForTermination() {
+	for {
+		select {
+		case <-assistant.done:
+			return
+		case <-assistant.ctx.Done():
+			assistant.Stop()
+		default:
+			continue
```

Comment:
Basically the way that defaults work, is they are the case that is run whenever no other case is ready. However, when you do something like this, where the default is just continue basically what's happening is, the for loop enters the select statement, none of the first two cases are done, so it goes to the default, which continues, thus starting another loop of the for loop, and this continues infinitely. Thus, this code never actually blocks and is just a for loop running infinitely. This is a massive memory hog and shouldn't be done. If you don't actually have something to do in a default case, then don't include one at all. If there's no default case, then this for loop would block instead of looping infinitely.